### PR TITLE
feat(): Add trailing newline when saving package.json

### LINKED ISSUE #4049: TNS Commands Removing Newline Added by NPM from Package.json
**Environment**
 - CLI: 4.2.4
 - Cross-platform modules: 4.3.0-2018-08-10-02
 - Android Runtime: 5.0.0-2018-10-18-162704-01
 - iOS Runtime: 4.3.0-2018-09-28-125108-01
 - Plugin(s): None

**Describe the bug**
By default NPM adds a \n newline to the end of the package.json file when npm saves the package.json file. 
[The line in npm-cli that calls the module that adds the newline](https://github.com/npm/cli/blob/14bd214c9e3fb42005dc1925d4e2a8ebe95a87f7/lib/install/save.js#L127)
[The module, maintained by npm, that adds the newline (stringify-package)](https://github.com/npm/stringify-package/blob/d8ad1ee57190db2a94b5915f0efd06a64639c2ac/index.js#L16)

Since the nativescript-cli doesn't use the same stringify command within FileSystem.writeJson it removes the newline added by npm-cli commands. This causes package.json to appear modified even when nothing has changed which makes for unneeded commits. For reference, below is the writeJson method within nativescript-cli:
https://github.com/NativeScript/nativescript-cli/blob/616782000380e5d3e96e3f013794644d58de2976/lib/common/file-system.ts#L208

**To Reproduce**
`tns create MyApp`
`cd ./MyApp`
The package.json file will have an extra newline
`tns platform update android`
The package.json file will NOT have an extra newline

**Expected behavior**
tns operations should not remove newlines from the end of package.json

**Additional context**
I am happy to open a pull request to modify the writeJson method so that it uses the stringify-package module (the same one that NPM uses and maintains).  I would make it add the newline if the file that is being saved is called package.json only. This way it won't affect any other usage of writeJson. I'd like to discuss this here and will open the pull request if everyone agrees.

### PR DESCRIPTION
**Pull request by Logikgate**
## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [x] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [x] You have signed the [CLA](http://www.nativescript.org/cla).
- [x] All existing tests are passing: https://github.com/NativeScript/nativescript-cli/blob/master/CONTRIBUTING.md#contribute-to-the-code-base
- [ ] Tests for the changes are included.

## What is the current behavior?
Nativescript-CLI commands that save package.json are stripping the trailing newline that NPM inserts when saving package.json. This can cause un-needed commits of package.json where the only thing that has changed is the removal of the trailing newline.

## What is the new behavior?
This pull request uses the module stringify-package, the same one used by NPM, to stringify package.json so that the formatting is kept the same.  The method writeJSON now uses stringify-package only if it is saving a file named package.json.

Fixes/Implements/Closes #4049 .

### CODE REVIEW

File: lib/common/file-system.ts [hunk 1]
```diff
@@ -205,7 +207,13 @@ export class FileSystem implements IFileSystem {
 			space = this.getIndentationCharacter(filename);
 		}
 
-		return this.writeFile(filename, JSON.stringify(data, null, space), encoding);
+		let stringifiedData = JSON.stringify(data, null, space);
```
Comment (KristianDD) [inline lib/common/file-system.ts]: Maybe this should be moved inside an `else` statement. No need to stringify data multiple times.
```TypeScript
let stringifiedData;
if (path.basename(filename) === "package.json") {
	const newline = detectNewline(stringifiedData);
	stringifiedData = stringifyPackage(data, space, newline);
} else {
	stringifiedData = JSON.stringify(data, null, space);
}
```

File: lib/common/file-system.ts [hunk 2]
```diff
@@ -205,7 +207,13 @@ export class FileSystem implements IFileSystem {
...
+		if (path.basename(filename) === "package.json") {
```
Comment (KristianDD) [inline lib/common/file-system.ts]: `package.json` might be used from `constants`:
```
import { PACKAGE_JSON_FILE_NAME } from "../constants";
if (path.basename(filename) === PACKAGE_JSON_FILE_NAME ) {
...
}
```

File: package.json [hunk 3]
```diff
@@ -37,6 +37,7 @@
     "cli-table": "https://github.com/telerik/cli-table/tarball/v0.3.1.2",
     "color": "3.0.0",
     "colors": "1.1.2",
+    "detect-newline": "^2.1.0",
```
Comment (KristianDD) [inline package.json]: It is safer to use an exact version. You can use `npm i detect-newline --save-exact`. Same goes for `stringify-package` dependency.

**Comment by radeva** [comment]
Hi @Logikgate ,

Congratulations on being one of the winners in the {N} First-time contributors contest! 

You can claim your prize by contacting us at nativescriptwinners[at]progress.com not later than Nov 30th 2018 . Make sure you send us the following info: 
👉Your full name 
👉Your email 
👉Your country or residence 

Best regards,
The NativeScript Team

**Pull request merged**